experimental/rank1_bnns/refining.py

```diff
@@ -0,0 +1,119 @@
+import numpy as np
+import tensorflow.compat.v2 as tf
+import tensorflow_probability as tfp
+
+from experimental.rank1_bnns.rank1_bnn_layers import DenseRank1, Conv2DRank1
+
+
+def get_auxiliary_posterior(posterior_mean,
+                            posterior_scale,
+                            prior_mean,
+                            prior_scale,
+                            auxiliary_scale):
+  """ Calculate the posterior distribution of an additive Gaussian """
+  """ auxiliary variable. q(a)=\int p(a|w)q(w)dw. """
+  prior_var = tf.math.pow(prior_scale, 2)
+  posterior_var = tf.math.pow(posterior_scale, 2)
+  auxiliary_var = tf.math.pow(auxiliary_scale, 2)
+  auxiliary_posterior_mean = (posterior_mean - prior_mean) \
+      * auxiliary_var / prior_var
+  auxiliary_posterior_var = posterior_var * tf.math.pow(auxiliary_var, 2) \
+      / tf.math.pow(prior_var, 2) + auxiliary_var * \
+      (prior_var - auxiliary_var) / prior_var
+  return auxiliary_posterior_mean, tf.sqrt(auxiliary_posterior_var)
+
+
+def get_conditional_prior(prior_mean,
+                          prior_scale,
+                          auxiliary_scale,
+                          auxiliary_sample):
+  """ Calculate the conditional prior given the value of an additive """
+  """ Gaussian auxiliary variable. p(w|a). """
+  prior_var = tf.math.pow(prior_scale, 2)
+  auxiliary_var = tf.math.pow(auxiliary_scale, 2)
+  return prior_mean + auxiliary_sample, tf.sqrt(prior_var - auxiliary_var)
+
+
+def get_conditional_posterior(posterior_mean,
+                              posterior_scale,
+                              prior_mean,
+                              prior_scale,
+                              auxiliary_scale,
+                              auxiliary_sample):
+  """ Calculate the conditional posterior given the value of an additive """
+  """ Gaussian auxiliary variable. q(w|a)\propto p(a|w)q(w). """
+  prior_var = tf.math.pow(prior_scale, 2)
+  posterior_var = tf.math.pow(posterior_scale, 2)
+  auxiliary_var = tf.math.pow(auxiliary_scale, 2)
+  conditional_mean = (prior_mean + (auxiliary_sample * posterior_var *
+                                    prior_var + (posterior_mean - prior_mean)
+                                    * (prior_var - auxiliary_var)
+                                    * prior_var)
+                      / (posterior_var * auxiliary_var + prior_var *
+                         (prior_var - auxiliary_var)))
+  conditional_var = posterior_var * prior_var * (prior_var - auxiliary_var) / \
+                    (auxiliary_var * posterior_var + prior_var *
+                     (prior_var - auxiliary_var))
+  return conditional_mean, tf.sqrt(conditional_var)
+
+
+def sample_rank1_auxiliaries(model, auxiliary_var_ratio):
+  for layer in model.layers:
+    if isinstance(layer, DenseRank1) or isinstance(layer, Conv2DRank1):
+      for rv_name, rv, regularizer in [('alpha', layer.alpha,
+                                        layer.alpha_regularizer),
```

Comment:
I would use `layer.alpha_initializer.mean` and `layer.alpha_initializer.stddev_constraint(layer.alpha_initializer.stddev)` (likewise for gamma), similar to what Dustin suggested below. The `layer.alpha_initializer.stddev` variable is an unconstrained stddev.

Reply:
Great idea!